src/dist/replication/meta_server/meta_bulk_load_service.cpp

```diff
@@ -497,6 +561,170 @@ void bulk_load_service::handle_app_unavailable(int32_t app_id, const std::string
+                pid,
+                pinfo.metadata.files.size(),
+                pinfo.metadata.file_total_size);
+        });
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::update_partition_status_on_remote_storage(const std::string &app_name,
+                                                                  const gpid &pid,
+                                                                  bulk_load_status::type new_status,
+                                                                  bool should_send_request)
+{
+    zauto_read_lock l(_lock);
+    partition_bulk_load_info pinfo = _partition_bulk_load_info[pid];
+
+    if (pinfo.status == new_status && new_status != bulk_load_status::BLS_DOWNLOADING) {
+        dinfo_f("app({}) partition({}) old status:{} VS new status:{}, ignore it",
+                app_name,
+                pid,
+                dsn::enum_to_string(pinfo.status),
+                dsn::enum_to_string(new_status));
+        return;
+    }
+
+    if (_partitions_pending_sync_flag[pid]) {
+        ddebug_f("app({}) partition({}) has already sync bulk load status, wait for next round",
+                 app_name,
+                 pid);
+        return;
+    }
+
+    _partitions_pending_sync_flag[pid] = true;
+    pinfo.status = new_status;
+    blob value = json::json_forwarder<partition_bulk_load_info>::encode(pinfo);
+
+    _meta_svc->get_meta_storage()->set_data(
+        get_partition_bulk_load_path(pid),
+        std::move(value),
+        std::bind(&bulk_load_service::update_partition_status_on_remote_storage_reply,
+                  this,
+                  app_name,
+                  pid,
+                  new_status,
+                  should_send_request));
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::update_partition_status_on_remote_storage_reply(
+    const std::string &app_name,
+    const gpid &pid,
+    bulk_load_status::type new_status,
+    bool should_send_request)
+{
+    {
+        zauto_write_lock l(_lock);
+        auto old_status = _partition_bulk_load_info[pid].status;
+        _partition_bulk_load_info[pid].status = new_status;
+        _partitions_pending_sync_flag[pid] = false;
+
+        ddebug_f("app({}) update partition({}) status from {} to {}",
+                 app_name,
+                 pid,
+                 dsn::enum_to_string(old_status),
+                 dsn::enum_to_string(new_status));
+
+        if (new_status == bulk_load_status::BLS_DOWNLOADED && old_status != new_status) {
+            if (--_apps_in_progress_count[pid.get_app_id()] == 0) {
+                update_app_status_on_remote_storage_unlock(pid.get_app_id(), new_status);
+            }
+        }
+        // TODO(heyuchen): add other status
+    }
+    if (should_send_request) {
+        partition_bulk_load(app_name, pid);
+    }
+}
+
+// ThreadPool: THREAD_POOL_META_STATE
+void bulk_load_service::update_app_status_on_remote_storage_unlock(
```

Comment:
unlock is a verb. Call it `_no_lock` by convention.

Reply:
`unlocked` is OK